src/gpu/ffm/trainer.cu

```diff
@@ -0,0 +1,340 @@
+  CUDA_CHECK(cudaDeviceSetLimit(cudaLimitPrintfFifoSize, 0));
+  CUDA_CHECK(cudaDeviceSetCacheConfig(cudaFuncCachePreferL1));
+
+  this->model = new ModelGPU<T>(params);
+}
+
+template<typename T>
+Trainer<T>::Trainer(const T* weights, Params &params) : params(params), trainDataBatcher(params.nGpus) {
+  CUDA_CHECK(cudaSetDeviceFlags(cudaDeviceMapHost));
+  CUDA_CHECK(cudaSetDeviceFlags(cudaDeviceScheduleSpin));
+
+#if __CUDA_ARCH__ > 500
+  CUDA_CHECK(cudaDeviceSetLimit(cudaLimitDevRuntimeSyncDepth, 0));
+  CUDA_CHECK(cudaDeviceSetLimit(cudaLimitDevRuntimePendingLaunchCount, 0));
+#endif // CUDA 5.0
+  CUDA_CHECK(cudaDeviceSetLimit(cudaLimitStackSize, 0));
+  CUDA_CHECK(cudaDeviceSetLimit(cudaLimitPrintfFifoSize, 0));
+  CUDA_CHECK(cudaDeviceSetCacheConfig(cudaFuncCachePreferL1));
+
+  this->model = new ModelGPU<T>(params, weights);
+}
+
+template<typename T>
+void Trainer<T>::setDataset(const Dataset<T> &dataset) {
+  DatasetBatcherGPU<T> *batcher = new DatasetBatcherGPU<T>(dataset, params);
+  trainDataBatcher[0] = batcher;
+}
+
+template<typename T>
+Trainer<T>::~Trainer() {
+  delete trainDataBatcher[0];
+  delete model;
+  CUDA_CHECK(cudaDeviceReset());
+}
+
+__constant__ int cK[1];
+__constant__ int cMaxRowSize[1];
+__constant__ int cRows[1];
+__constant__ int cAlignFeat[2];
+__constant__ float cRegLambda[1];
+__constant__ float cLearningRate[1];
+__constant__ int cWeightsOffset[1];
+__constant__ int cBatchOffset[1];
+
+
+template<typename T>
+__global__ void wTxKernel(const int *__restrict__ features, const int *__restrict__ fields, const T *__restrict__ values,
+                          const T *__restrict__ r, const int *__restrict__ rowSizes,
+                          T *__restrict__ weightsPtr, T *__restrict__ losses, const bool update, const int *__restrict__ labels = nullptr) {
+  int rowIdx = (blockIdx.x * blockDim.x + threadIdx.x) / cMaxRowSize[0];
+
+  if(rowIdx >= cRows[0]) return;
+
+  int rowSize = rowSizes[rowIdx + 1] - rowSizes[rowIdx];
+
+  int nodeIdx = (blockIdx.x * blockDim.x + threadIdx.x) % cMaxRowSize[0];
+
+  if(nodeIdx >= rowSize) return;
+
+  __shared__ int fieldFeature[MAX_BLOCK_THREADS * 2];
+  __shared__ T vals[MAX_BLOCK_THREADS];
+  __shared__ T scales[MAX_BLOCK_THREADS];
+
+  __shared__ T kappas[MAX_BLOCK_THREADS];
+  __shared__ T expnyts[MAX_BLOCK_THREADS];
+
+  int n1 = rowSizes[rowIdx] + nodeIdx - cBatchOffset[0];
+
+  fieldFeature[threadIdx.x * 2] = fields[n1];
+  fieldFeature[threadIdx.x * 2 + 1] = features[n1];
+  vals[threadIdx.x] = values[n1];
+  scales[rowIdx % MAX_BLOCK_THREADS] = r[rowIdx];
+
+  if(update) {
+    expnyts[rowIdx % MAX_BLOCK_THREADS] = std::exp(-labels[rowIdx] * losses[rowIdx]);
+    kappas[rowIdx % MAX_BLOCK_THREADS] = -labels[rowIdx] * expnyts[rowIdx % MAX_BLOCK_THREADS] / (1 + expnyts[rowIdx % MAX_BLOCK_THREADS]);
+  }
+
+  __syncthreads();
+
+  T loss = 0.0;
+
+  for(int i = 1; n1 + i < rowSizes[rowIdx + 1] - cBatchOffset[0]; i++) {
```

Comment:
Not sure, instead of spinning up a thread for each field:feature:value (lets call it a node) tuple and running this loop for each consecutive node in that row we could spin up a thread for each node pair. Didn't try that approach yet as I was afraid of thread overhead.

Currently the main slowdown I'm noticing is becuase of:

1) number of registers used by the `wTx` kernel which limits the number of concurrent blocks being ran

2) read/write from global memory when using the weights array since that access isn't very coalesce within blocks. Interesting experiment is to move lines 157/158:

```
(weightsPtr + idx1)[d+1] = w1gdup;
(weightsPtr + idx2)[d+1] = w2gdup;
```

Right after:

```
const T w1gdup = (weightsPtr + idx1)[d+1] + g1 * g1;
const T w2gdup = (weightsPtr + idx2)[d+1] + g2 * g2;
```

There's a visible (~10%) slowdown on relatively large data (400k rows, 39 nodes in each row). I'm assuming this is due to how CUDA loads/stores data.

Basically the major slowdown in coming from the `if (update) {` branch. Putting both weights and gradients into the same array helped quite a bit but even on 1080Ti this is only ~2-3x faster than the CPU implementation.